### ISSUE TITLE
Path Operators "%" error

### ISSUE THREAD

**Issue by wangdaowangdao** [IBM/JSONata4Java#214]
when use the "%" path operator as follows, 

Account.Order.Product.{
 'Product': `Product Name`,
 'Order': %.OrderID,
 'Account': %.%.`Account Name`
}

encounter error as follows, 

line 3:10 at [@13,62:62='%',<31>,3:10]: mismatched input '%' expecting {'(', 'true', 'false', STRING, 'null', '[', '{', '$', '$$', '**', NUMBER, FUNCTIONID, '*', '-', VAR_ID, ID} , line 3:11 at [@14,63:63='.',<1>,3:11]: extraneous input '.' expecting {'(', 'true', 'false', STRING, 'null', '[', '{', '$', '$$', '**', NUMBER, FUNCTIONID, '*', '-', VAR_ID, ID} , line 3:19 at [@16,71:71=',',<8>,3:19]: mismatched input ',' expecting {<EOF>, '.', 'in', '?', '[', '{', '~>', '*', '/', '+', '-', '%', '=', '!=', '<', '<=', '>', '>=', '&', 'and', 'or'}

use the test ui for test also encounter error,

JSONata syntax error: line 3:10 at [@13,62:62='%',<34>,3:10]: mismatched input '%' expecting {'(', 'true', 'false', STRING, 'null', '[', '{', '$', '$$', '**', NUMBER, FUNCTIONID, '*', '/', '-', VAR_ID, ID}
, line 3:11 at [@14,63:63='.',<1>,3:11]: extraneous input '.' expecting {'(', 'true', 'false', STRING, 'null', '[', '{', '$', '$$', '**', NUMBER, FUNCTIONID, '*', '/', '-', VAR_ID, ID}
, line 3:19 at [@16,71:71=',',<8>,3:19]: mismatched input ',' expecting {<EOF>, '.', 'in', '?', '[', '{', '~>', '*', '/', '+', '-', '%', '=', '!=', '<', '<=', '>', '>=', '&', 'and', 'or'}

the other path operator seems also have errors.

**Comment by jiesiro** [comment]
me to

**Comment by wnm3** [comment]
We don't support this operator at this time due to the parser approach we have taken. It differs from the JavaScript parser.

**Comment by wangdaowangdao** [comment]
Got it. Thanks. Is there any plan to support this operator in the future?

**Comment by bluemel** [comment]
On new branch "241-path-operator" I've made a little first experiment with the parser to support the "%" operator.
Simply extended the "expr" in MappingExpression.g4 by line
| '%'     # parent_path
I am not perfectly sure if this is the way to go but we have so far
- no syntax error anymore throw by the parser
- generated a new method "public JsonNode visitParent_path(Parent_pathContext ctx)" we coud implement in class ExpressionsVisitor
Anyone who is interested might check out this branch and run my new unit test class
com.api.jsonata4java.test.expressions.path.PathErrorTests

What do you think?
Do we have a chance with this approach?
How to evaluate the mapping expression containing parent path expressions correctly?

**Comment by wnm3** [comment]
Thank you -- the issue is having to create a parser for the current data to work down to the current location, maintaining the path of parents used to arrive there in order to have the necessary data to reference. If you have time to experiment further that would be great.

**Comment by bluemel** [comment]
So in my eyes, the parser part is already in place on branch 214-path-operator.
What is missing is the implementation of an according expression evaluation.
Here I lack some understanding how this evaluation works in general and where we could start to hang in the implementation.
Probably it is somewhere in the ExpressionVisitor. When I implement a method visitParent_path(Parent_pathContext ctx) ther, I can see that it is already called, when I run my unit test (see class PathErrorTests).
But I do not know what I could do there and if this method is and appropriate point to come in with some implementation.
Do you have an idea how we could make the path resolution working out?

**Comment by wnm3** [comment]
Yes, we'd simply have to fill out the logic in the new method in the ExpressionVisitor. I suppose the easiest way to understand what is required is to just try to write a function that can parse a JSON object from its root keys until it find the object you've chosen whose parents you want. You are essentially navigating a tree with its root at the entire object level, and working through the keys recursively to find the object, filling a stack with the parents and popping them when you reach a leaf without finding what you want. Once you find it, you can use the stack you've accumulated to get the parent and the grandparent, etc. Once we have this logic then we can try to make it work with the current context, etc. Each of the methods in the ExpressionVisitor is handed a context variable with the tree of parsed parts for the expression.

Probably just try writing the code for the JSON Object traversal after you've initially picked an object it contains to see if you can find the path that leads to it.

**Comment by wnm3** [comment]
This might save time: https://stackoverflow.com/questions/54421810/how-to-get-the-parent-element-name-of-a-child-element-in-json

**Comment by Honeylemonade** [comment]
if this project support % will be excellent!

**Comment by jiesiro** [comment]
> 在新分支“241-path-operator”上，我用解析器做了第一个小实验，以支持“%”运算符。 简单地扩展 MappingExpression.g4 中的“expr”行 | '%' # parent_path 我不完全确定这是否是要走的路，但我们到目前为止
> 
> * 解析器不再抛出语法错误
> * 生成了一个新方法“public JsonNode visitParent_path(Parent_pathContext ctx)”，我们可以在 ExpressionsVisitor 类中实现
> 
> 任何有兴趣的人都可以查看这个分支并运行我的新单元测试类 com.api.jsonata4java.test.expressions.path.PathErrorTests
> 
> 你怎么看？ 我们有机会采用这种方法吗？ 如何正确评估包含父路径表达式的映射表达式？

I wanted to experiment with this branch but there seems to be a lack of jar packages 
<img width="1160" alt="image" src="https://user-images.githubusercontent.com/55075594/197755930-4e3bc672-3127-47e4-ae5e-b6e511fe0e8a.png">

**Comment by bluemel** [comment]
Hi Jesiro, THX for your support. Running the Maven build like Nat described you should find the missing Java source files generated into folder "target/generated-sources/antlr4".

**Comment by bluemel** [comment]
By the way... I've recently achieved already a partial success and just committed it on that branch.

The new ExpressionsVisitor method visitParent_path() now
- takes the top level context JsonNode from the (_environment.) stack
- builds up (lazily) a "parent node map"
- and if it finds the parent tries to resolve the rest of the path

This works out fine now for %.<path> but not for %.%.<path>.
This is probably a parser issue.

**Comment by bluemel** [comment]
OK guys. It seems to me that we can also cope now with an arbtrary sequence of parent path operators.

ANTLR grammar definitions line:
| (('%' '.')+ expr)    # parent_path
and a little extension in the expression evaluation code seem to do the job.

So please guys... test the stuff now very very toroughly before we merge this hack.
@jiesiro you're very welcomed to test the new code on the branch e. g. by using JSONata4Java's new Tester UI.
@wnm3 Nat:
- would you please review very toroughly what I did so far, and tell me your opinion about the solution I've chosen?
  Is the approach generalistic enough?
- could you lead me to test / test grups in the agnostic test suite that might have been excluded so far a can be included now?
I'll also try to test this with further automated tests.

**Comment by bluemel** [comment]
Recently fixed the algorithm.

**Comment by bluemel** [comment]
Basic questions:
- how is the PathExpression part of the ANTLR (PathExpressionLexer.g4, PathExpressionParser.g4) grammar involved into evaluation of paths? Did I do right not to touch this part of the grammar at all, when trying to implement evaluation of '%'?
- how do the PathExpressionTests work? Particularly... what is "schemaJsonString" (weird grammar) and "valueToAssignJsonString"?... I have not the least idea

**Comment by bluemel** [comment]
@jiesiro Jiesiro, did you succeed in running a Maven buildfrom command line or from the IDE of your choice (IntellyJ?) or do you need help to set up a build environment? I would appreciate help in testing what I did so far.

**Comment by jiesiro** [comment]
> @jiesiroJiesiro，您是从命令行还是从您选择的 IDE（IntellyJ？）成功运行 Maven 构建，还是需要帮助来设置构建环境？我会感谢帮助测试我到目前为止所做的事情。

I'm sorry, I just saw that, that's great, I've tested it very well, thank you very much, when do I add the main branch

**Comment by jiesiro** [comment]
If the father is an array and the code doesn't match by way of [0] maybe that's a problem? 
<img width="901" alt="image" src="https://user-images.githubusercontent.com/55075594/198018289-7e204a45-23ce-4ade-baba-8680952b2f1e.png">
json:{
    "status": 0,
    "message": "OK",
    "data": [
        {
            "product": "ipcd",
            "namespace": "carr2_amps_%.pv",
            "tags": [
                "module=yq01-d2","module=yq01-d3"
            ],
            "metrics": [
                {
                    "metric": "value",
                    "histoCount": 1,
                    "history": [
                        {
                            "value": 0.0,
                            "timestamp": 1660046403
                        },
                        {
                            "value": 0.0,
                            "timestamp": 1660046414
                        }
                  
                    ]
                }
            ]
        }
    ]
}
jsonata:

data.metrics.history.{
       "objectId":%.%.tags[0],
       "test":"test"
}
java:
<img width="1032" alt="image" src="https://user-images.githubusercontent.com/55075594/198018609-1b72076a-5608-4abe-91f6-83e6cf0c69b8.png">
"objectId" did not come out

**Comment by bluemel** [comment]
THX @jiesiro. Reproduced and fixed. Just update with git pull branch 214-path-operator again.

**Comment by wnm3** [comment]
@bluemel I don't believe I have used the PathExpressionLexer.g4 nor the PathExpressionParser.g4 -- only the ExpressionVisitor based on the classes produced by parsing the MappingExpression.g4 (and I should remove the MappingExpression.g4_new as that is obsolete).

The PathExpressionVisitor / PathExpression classes are used in the PathExpressionTests which are run as part of the check when building the jars but I think the tests are quite limited.

Regarding the schemaJsonString, etc., this work was originally started with the IoT (Internet of Things) and they had some additional well defined variables (e.g., $index) that could be set as part of their rule execution environment  The validation schemas were used to ensure proper types were being set. You'll see many of the tests that referenced $index have been commented out.

I think we would do better to add tests for the % Parent operator to BasicExpressionTests or create a test specific for this new capability and add it to the AgnosticTestSuite

**Comment by jiesiro** [comment]
@bluemel  
json:
{
"status": 0,
"message": "OK",
"data": [
{
"product": "ipcd",
"namespace": "carr2_amps_%.pv",
"tags": [
"module=yq01-d2","module=yq01-d3"
],
"metrics": [
{
"metric": "value",
"histoCount": 1,
"history": [
{
"value": 0.0,
"timestamp": 1660046403
},
{
"value": 0.0,
"timestamp": 1660046414
}

                ]
            }
        ]
    }
]}
jsonata :
($replaceSepStr:= $replace(?,"_%.","_")
~> $replace(?,"-.","_")
~> $replace(?,"-","_")
~> $replace(?,".","_")
~> $replace(?,"#","_")
~> $replace(?,":","_");

$addHeader:=$replace(?,"module=","")
~>$replace(?,"yq01-d2","YQ2_")
~>$replace(?,"yq01-d5","YQ5_");

data.metrics.history.{
        "modelId":"IDC_STAT",
        "objectId":$addHeader(%.%.tags[0])&''&$replaceSepStr(%.%.namespace,"_%.","_"),
        "bindName":"MAIN",
        "type":"PROPERTY",
        "timestamp":timestamp,
        "properties": {
            "value":value,
            "enterprise_id":"1",
            "device_instance_code":$addHeader(%.%.tags[0])&''&$replaceSepStr(%.%.namespace,"_%.","_")
            },
         "tags":[
            "device_instance_code",
            "enterprise_id"
            ]
})
java: 
<img width="1417" alt="image" src="https://user-images.githubusercontent.com/55075594/198185297-6811b03b-e0e7-4089-9478-e29651822661.png">
Is this nesting done? Or a bug? 
expected data:
<img width="1188" alt="image" src="https://user-images.githubusercontent.com/55075594/198186365-43fe151b-2fb1-4710-97d3-fa9cae9e58a2.png">

**Comment by wnm3** [comment]
I enabled the parent-operator tests in the AgnosticTestSuite.java but found the 27th test fails due to additional unsupported operators (the @$L reference of $L). You can see the target/jsonata-1.8.4/test/test-suite/groups/parent-operator/parent.json. So while leaving the test suite enabled by commenting out this line:
		//			"parent-operator", // issue #60
I had to add these lines to skip that particular test:
		// issue #60 parent-operator
		SKIP_CASES("parent-operator", "parent-27");
I've also set up the files to create release 2.1.3 assuming we are getting close.

**Comment by wnm3** [comment]
Go ahead and delete the MappingExpression.g4_new

Regarding the others, for now I'd leave them as they do form part of the test suite and aren't causing any harm -- just confusion. So adding something to the README.md would be fine in the area for development... So far you are the first person to be brave enough to alter the language ;^) so it hasn't been an issue to date.

**Comment by bluemel** [comment]
> ests in the AgnosticTestSuite.java but found the 27th test fails due to additional unsupported operator

Very good so far. But we should have some clarifications / fine tuning before we consider a new release (maybe 2.2.0? depends on our versioning policy / strategy e. g. semver / semantic versioning - this is a new feature probably maintaining full downward compatibility - no breaking change):

1) Error handling if the mapping expression contains parent operator exceeding the input JSON root - not yet implemented?

2) FrameEnvironment _enclosingFrame. What is this concept about? Can we imagine cases where it would be sensible to search for parent nodes there?

3) Performance: the current solution - indexing the whole JSON from the root context is "brute force" and might have performace impact with large input JSON. Could we imagine something smarter?

**Comment by wnm3** [comment]
For error handling, I'd follow the conventions from try.jsonata.org. e.g., create the problem and see what you get. For example, Account.% returns the entire object passed, whereas Account.%.% returns an exception stating:  
```
The object representing the 'parent' cannot be derived from this expression
```
see https://try.jsonata.org/kfHf_tZr7

Regarding the framing, there are situations where the focus of a subexpression is limited in scope to less than the original object. The enclosing frame is the parent of the current environment. I don't believe the parent operator should see beyond the current environment. An example is the parentheses code block. Variables created inside the parentheses only live until we exit the code block. Variables created in an environment replace those of the same name in the enclosing (parent) environment for the life of the code block. Afterward, the original variable value is available from the same named variable.

This works: https://try.jsonata.org/mfFL9JTkz However, if I add another parent operator, we get duplication in the result array. See https://try.jsonata.org/Swx5TEi00

These parent operators are kind of odd because you can't get the actual key of the array "object", it only seems to traverse the actual objects. See  https://try.jsonata.org/ryjA9t6fS that goes to the object keyed by foo even though with one less parent reference you get the object in the first cell of blah https://try.jsonata.org/BiOOUbffY

I think your inquiry is more about the context stack. It does not appear the parent operators work through the context stacks when limited using the $ operator.

Using the same input, ((foo.blah).baz.$)[0] works, but ((foo.blah).baz.$)[0].% has no parent, nor does (foo.blah).baz.$[0] have a parent, nor does foo.blah.baz.$[0], nor does foo.blah.baz.$

Regarding performance, though brute force, it seems to work. Perhaps you want to get some experience with it to see if someone has a good example where performance is an issue?

**Comment by bluemel** [comment]
> I enabled the parent-operator tests in the AgnosticTestSuite.java but found the 27th test fails due to additional unsupported operators (the @$L reference of $L). You can see the target/jsonata-1.8.4/test/test-suite/groups/parent-operator/parent.json. So while leaving the test suite enabled by commenting out this line: // "parent-operator", // issue #60 I had to add these lines to skip that particular test: // issue #60 parent-operator SKIP_CASES("parent-operator", "parent-27"); I've also set up the files to create release 2.1.3 assuming we are getting close. I also updated the pom.xml to address newer library versions.

- As already said: I'd propose to have new version 2.2.0 according to the "Semantic Versioning" standard (see https://semver.org/).
- 27 tests successfull out of 28 is not that bad... however simple expressions like a standalone '%' still fail :-( see also https://try.jsonata.org/9LAhagQCC - cool sharing feature by the way 
 
So 

> @bluemel json: { "status": 0, "message": "OK", "data": [ { "product": "ipcd", "namespace": "carr2_amps_%.pv", "tags": [ "module=yq01-d2","module=yq01-d3" ], "metrics": [ { "metric": "value", "histoCount": 1, "history": [ { "value": 0.0, "timestamp": 1660046403 }, { "value": 0.0, "timestamp": 1660046414 }
> 
> ```
>             ]
>         }
>     ]
> }
> ```
> 
> ]} jsonata : ($replaceSepStr:= $replace(?,"_%.","_") ~> $replace(?,"-.","_") ~> $replace(?,"-","_") ~> $replace(?,".","_") ~> $replace(?,"#","_") ~> $replace(?,":","_");
> 
> $addHeader:=$replace(?,"module=","") ~>$replace(?,"yq01-d2","YQ2_") ~>$replace(?,"yq01-d5","YQ5_");
> 
> data.metrics.history.{ "modelId":"IDC_STAT", "objectId":$addHeader(%.%.tags[0])&''&$replaceSepStr(%.%.namespace,"_%.","_"), "bindName":"MAIN", "type":"PROPERTY", "timestamp":timestamp, "properties": { "value":value, "enterprise_id":"1", "device_instance_code":$addHeader(%.%.tags[0])&''&$replaceSepStr(%.%.namespace,"_%.","_") }, "tags":[ "device_instance_code", "enterprise_id" ] }) java: <img alt="image" width="1417" src="https://user-images.githubusercontent.com/55075594/198185297-6811b03b-e0e7-4089-9478-e29651822661.png"> Is this nesting done? Or a bug? expected data: <img alt="image" width="1188" src="https://user-images.githubusercontent.com/55075594/198186365-43fe151b-2fb1-4710-97d3-fa9cae9e58a2.png"> By the way, when do you expect to join the master branch?

Reproduced (see also https://try.jsonata.org/sAasmk6un).
Need some time for analysis.

**Comment by bluemel** [comment]
> @bluemel json: { "status": 0, "message": "OK", "data": [ { "product": "ipcd", "namespace": "carr2_amps_%.pv", "tags": [ "module=yq01-d2","module=yq01-d3" ], "metrics": [ { "metric": "value", "histoCount": 1, "history": [ { "value": 0.0, "timestamp": 1660046403 }, { "value": 0.0, "timestamp": 1660046414 }
> 
> ```
>             ]
>         }
>     ]
> }
> ```
> 
> ]} jsonata : ($replaceSepStr:= $replace(?,"_%.","_") ~> $replace(?,"-.","_") ~> $replace(?,"-","_") ~> $replace(?,".","_") ~> $replace(?,"#","_") ~> $replace(?,":","_");
> 
> $addHeader:=$replace(?,"module=","") ~>$replace(?,"yq01-d2","YQ2_") ~>$replace(?,"yq01-d5","YQ5_");
> 
> data.metrics.history.{ "modelId":"IDC_STAT", "objectId":$addHeader(%.%.tags[0])&''&$replaceSepStr(%.%.namespace,"_%.","_"), "bindName":"MAIN", "type":"PROPERTY", "timestamp":timestamp, "properties": { "value":value, "enterprise_id":"1", "device_instance_code":$addHeader(%.%.tags[0])&''&$replaceSepStr(%.%.namespace,"_%.","_") }, "tags":[ "device_instance_code", "enterprise_id" ] }) java: <img alt="image" width="1417" src="https://user-images.githubusercontent.com/55075594/198185297-6811b03b-e0e7-4089-9478-e29651822661.png"> Is this nesting done? Or a bug? expected data: <img alt="image" width="1188" src="https://user-images.githubusercontent.com/55075594/198186365-43fe151b-2fb1-4710-97d3-fa9cae9e58a2.png"> By the way, when do you expect to join the master branch?

Reproduced (see also https://try.jsonata.org/sAasmk6un).
Need some time for analysis and hopefully find fixes or at least document restrictions if we have to leave them.

**Comment by jiesiro** [comment]
> @jiesiro谢谢你的测试。当您在 try.jsonata.org 网站上创建示例时，您能否选择分享您的设置，以便我们可以从该网站复制 JSON / 表达式？
> 
> 当您单击 右上角菜单中的共享图标时，它会提供此对话框： 它会为您提供一个引用您的设置的 URL，您可以在此处粘贴问题。 ![图片](https://user-images.githubusercontent.com/1359024/198302430-f21778b2-7e50-4baa-b323-13b1ae7eb6b2.png) ![图片](https://user-images.githubusercontent.com/1359024/198302176-a88e3e01-45a7-47a0-b13c-39a1db846192.png)
> 
> 谢谢你。

Well, thanks for reminding me, I found that %.% is already available for the master branch. The path operator, the path operator has crossed into the main branch, right?

**Comment by bluemel** [comment]
So far as I know and see % support is not yet merged to master.
See https://github.com/IBM/JSONata4Java/network
If you don't mind you could help yourself for the next days by producing your own "pre release" version 2.2.0-SNAPSHOT from branch 214-path-operator.

**Comment by jiesiro** [comment]
> 



> 我的预表是你的https://“你自己的 支持[还](https://github.com/IBM/JSONata4Java/network)没有合并到中。”版本 2.2.0-SNAPSHOT 来帮助您自己。
<img width="1186" alt="image" src="https://user-images.githubusercontent.com/55075594/198523749-984dba96-f04a-4046-8305-6ceb3e655efc.png">

I think so too, but the master branch I'm currently using runs %.%. Path operator and %.%.[0] without any problems

**Comment by bluemel** [comment]
Not on my machine... @jiesiro, are you sure you've switched back to the main branch  a n d  built the project again?

**Comment by bluemel** [comment]
Fixed the solitary % edge case. ANTLR does a good job here.

**Comment by bluemel** [comment]
@jiesiro, concerning your example https://try.jsonata.org/sAasmk6un

It is failing under JSONata4Java not because of any issue with '%' but because JSONata for Java does not yet support other language constructs you've applied:
- function chain operator
- this kind of function definition without keyword function using ? as argument placeholder - how is this called and where is it documented in JSONata?

If you workaround without the chain operator and define functions with keyword "function" you should get a result faily close to what you've requested.
see my workaround https://try.jsonata.org/CXvQ2Lg9e

**Comment by jiesiro** [comment]
> @jiesiro，关于您的示例https://try.jsonata.org/sAasmk6un
> 
> 它在 JSONata4Java 下失败不是因为 '%' 的任何问题，而是因为 JSONata for Java 尚不支持您应用的其他语言结构：
> 
> * 函数链算子
> * 没有关键字函数的函数定义使用？作为参数占位符 - 这是如何调用的，它在 JSONata 中记录在哪里？
> 
> 如果您在不使用链表结果的结果的情况下工作并使用“函数”定义您的任务，您应该得到接近结果的方法[。https:](https://try.jsonata.org/CXvQ2Lg9e) //try.jsonorg/CXvQ2Lg9e[](https://try.jsonata.org/CXvQ2Lg9e)

I guess it came in during the run. Also, this suggestion doesn't seem to be supported in java,   get an error. I hope it can be processed and merged into the main branch as soon as possible, because I am really in a hurry to use it, thank you very much

**Comment by wnm3** [comment]
Moving to 2.2.0 is fine with me. I shall update the components in this branch to prepare for merging and creating 2.2.0

**Comment by bluemel** [comment]
Ok Nat. I'd personally feel better if we had more test... bit... in order to make our useres happy, I'm also fine with a quick release.
Some points maybe:
- I just now added a primitive error handling in case % is applied to often.
  Because lacking better ideas, I've just thrown a RuntimeException. Seems to work... but I'd prefer to thow a distinct one. EvaluateException unfortunately is a "checked" one.
- I added an inclusion of every JUnit test ending with"*Test" (not "*Tests) in order to be able to write additionally simple unit test  s - which somtimes can held a lot. In the long run we should think about a better test concept... the current one is not optimal.
Happy release
